migration/idempotent/017-metric-rollup.sql

```diff
@@ -0,0 +1,290 @@
+-- 1. Delete the entries in _prom_catalog.rollup
+-- 2. Delete the entries in _prom_catalog.metric_with_rollup
+-- 3. Delete the actual Cagg views that represent the rollups. This is done by directing deleting the entire rollup schema
+CREATE OR REPLACE PROCEDURE _prom_catalog.delete_metric_rollup(rollup_name TEXT) AS
+$$
+    DECLARE
+        rollup_schema_name TEXT;
+
+    BEGIN
+        EXECUTE FORMAT('SELECT schema_name FROM _prom_catalog.rollup WHERE name = %L', rollup_name) INTO rollup_schema_name;
+        IF rollup_schema_name IS NULL THEN
+            RAISE EXCEPTION 'Cannot delete metric rollup % due to missing schema', rollup_name;
+        END IF;
+        EXECUTE FORMAT('DELETE FROM _prom_catalog.rollup WHERE name = %L', rollup_name);
+        EXECUTE FORMAT('DELETE FROM _prom_catalog.metric_with_rollup WHERE rollup_schema = %L', rollup_schema_name);
+        EXECUTE FORMAT('DROP SCHEMA %I CASCADE', rollup_schema_name);
+    END;
+$$
+LANGUAGE PLPGSQL;
+
+-- scan_for_new_rollups is called in regular intervals to scan for either new metrics or for new resolution
+-- and create metric rollups for them.
+CREATE OR REPLACE PROCEDURE _prom_catalog.scan_for_new_rollups(job_id int, config jsonb) AS
+$$
+DECLARE
+    r RECORD;
+    m RECORD;
+    new_rollups_created INTEGER := 0;
+    rollup_view_created BOOLEAN;
+
+BEGIN
+    IF ( SELECT prom_api.get_automatic_downsample()::BOOLEAN IS FALSE ) THEN
+        RETURN;
+    END IF;
+
+    FOR r IN
+        SELECT * FROM _prom_catalog.rollup
+    LOOP
+        rollup_view_created := FALSE;
+        new_rollups_created := 0;
+        FOR m IN
+            EXECUTE FORMAT('select metric_name, table_name FROM _prom_catalog.metric where metric_name NOT IN (
+                SELECT metric_name FROM _prom_catalog.metric_with_rollup WHERE rollup_schema = %L
+            )', r.schema_name) -- Get metric names that have a pending rollup creation.
+        LOOP
+            SELECT INTO rollup_view_created _prom_catalog.create_metric_rollup_view(r.schema_name, m.metric_name, m.table_name, r.resolution);
+            IF rollup_view_created THEN
+                EXECUTE FORMAT('INSERT INTO _prom_catalog.metric_with_rollup VALUES (%L, %L, %L, TRUE)', r.schema_name, m.metric_name, m.table_name);
+                new_rollups_created := new_rollups_created + 1;
+            END IF;
+        END LOOP;
+        RAISE LOG '[Rollup] Created % for rollup name %', new_rollups_created, r.name;
+    END LOOP;
+
+    COMMIT;
+
+    -- Refresh the newly added rollups for entire duration of data.
+    FOR r IN
+        SELECT rollup_schema, table_name FROM _prom_catalog.metric_with_rollup WHERE refresh_pending = TRUE
+    LOOP
+        CALL refresh_continuous_aggregate(r.rollup_schema || '.' || r.table_name, NULL, NULL);
+        UPDATE _prom_catalog.metric_with_rollup SET refresh_pending = FALSE WHERE rollup_schema = r.rollup_schema AND table_name = r.table_name;
```

Comment:
Yes, I was thinking in that direction when I was programing this. My reasoning was that this will do frequent flushing (so flush batch sizes won't be huge) and avoid too much buffering in memory. Let's do `COMMIT` after every refresh.